Sources/AblyChat/Messages.swift

```diff
@@ -0,0 +1,74 @@
+import Ably
+
+public protocol Messages: AnyObject, Sendable, EmitsDiscontinuities {
+    func subscribe(bufferingPolicy: BufferingPolicy) -> MessageSubscription
+    func get(options: QueryOptions) async throws -> any PaginatedResult<Message>
+    func send(params: SendMessageParams) async throws -> Message
+    var channel: ARTRealtimeChannelProtocol { get }
+}
+
+public struct SendMessageParams: Sendable {
+    public var text: String
+    public var metadata: MessageMetadata?
+    public var headers: MessageHeaders?
+
+    public init(text: String, metadata: MessageMetadata? = nil, headers: MessageHeaders? = nil) {
+        self.text = text
+        self.metadata = metadata
+        self.headers = headers
+    }
+}
+
+public struct QueryOptions: Sendable {
+    public enum Direction: Sendable {
+        case forwards
+        case backwards
+    }
+
+    public var start: Date?
+    public var end: Date?
+    public var limit: Int?
+    public var direction: Direction?
+
+    public init(start: Date? = nil, end: Date? = nil, limit: Int? = nil, direction: QueryOptions.Direction? = nil) {
+        self.start = start
+        self.end = end
+        self.limit = limit
+        self.direction = direction
+    }
+}
+
+public struct QueryOptionsWithoutDirection: Sendable {
+    public var start: Date?
+    public var end: Date?
+    public var limit: Int?
+
+    public init(start: Date? = nil, end: Date? = nil, limit: Int? = nil) {
+        self.start = start
+        self.end = end
+        self.limit = limit
+    }
+}
+
+// Currently a copy-and-paste of `Subscription`; see notes on that one. For `MessageSubscription`, my intention is that the `BufferingPolicy` passed to `subscribe(bufferingPolicy:)` will also define what the `MessageSubscription` does with messages that are received _before_ the user starts iterating over the sequence (this buffering will allow us to implement the requirement that there be no discontinuity between the the last message returned by `getPreviousMessages` and the first element you get when you iterate).
+public struct MessageSubscription: Sendable, AsyncSequence {
+    public typealias Element = Message
+
+    public init<T: AsyncSequence>(mockAsyncSequence _: T) where T.Element == Element {
+        fatalError("Not yet implemented")
+    }
+
+    public func getPreviousMessages(params _: QueryOptionsWithoutDirection) async throws -> any PaginatedResult<Message> {
```

Comment:
Why not use `QueryOptions` ignoring direction?

Reply:
It came from [the use of `Omit<QueryOptions, 'direction'>`](https://github.com/ably/ably-chat-js/blob/0b4b0f8c35d1fb789bcc1fe779c920140f4b598c/src/core/messages.ts#L149) in the JS SDK (TypeScript has the ability to create an ad-hoc interface which is the same as another interface but with some properties removed; and there isn't an equivalent in Swift hence this clumsy new type).

We could instead use `QueryOptions` and document that the direction is ignored, yeah; I don't have a strong opinion.

Reply:
Also I've already asked elsewhere why this func even needed if there is another one in `Messages` which should return the same?

Reply:
The aim of this method ([here](https://github.com/ably/ably-chat-js/blob/0b4b0f8c35d1fb789bcc1fe779c920140f4b598c/src/core/messages.ts#L144-L149) is original in JS) is to address the following use case: the user subscribes to receive messages, and they start receiving messages. But they also want to display the messages that were sent before they subscribed (i.e. the message history). There's no easy way for them to do that using the `Messages` method that you mentioned, because they can’t form a `QueryOptions` that expresses "please give me all the messages that were sent before I created my subscription". The best they can do is try and guess an appropriate `end` query option, but that's going to be inaccurate and potentially lead to the message history either overlapping with the messages received through subscription, or some messages being missed in a gap between the two. The `MessageSubscription` method that you've asked about uses internal information about the state of the channel at the moment when the subscription was created in order to accurately form this history query.

Reply:
This is interesting how it solved internally, but if it can sew messages seamlessly it's very cool. Other day I just was opening socket and grabbing last N messages discarding all overlapping 😀